cmd/cluster/resize/infra_node.go

```diff
@@ -132,13 +156,62 @@ func (r *Resize) RunInfra(ctx context.Context) error {
 		return err
 	}
 
+	// Identify the original nodes and temp nodes
+	// requireInfra matches all infra nodes using the selector from above
+	requireInfra, err := labels.NewRequirement(infraNodeLabel, selection.Exists, nil)
+	if err != nil {
+		return err
+	}
+
+	// requireNotTempNode matches all nodes that do not have the temporaryInfraNodeLabel, created with the new (temporary) machine pool
+	requireNotTempNode, err := labels.NewRequirement(temporaryInfraNodeLabel, selection.DoesNotExist, nil)
+	if err != nil {
+		return err
+	}
+
+	// requireTempNode matches the opposite of above, all nodes that *do* have the temporaryInfraNodeLabel
+	requireTempNode, err := labels.NewRequirement(temporaryInfraNodeLabel, selection.Exists, nil)
+	if err != nil {
+		return err
+	}
+
+	// infraNode + notTempNode = original nodes
+	originalNodeSelector := selector.Add(*requireInfra, *requireNotTempNode)
+
+	// infraNode + tempNode = temp nodes
+	tempNodeSelector := selector.Add(*requireInfra, *requireTempNode)
+
+	// Drain the original nodes
+	// Possibly overkill since deleting the machinepool next should also trigger a drain
+	log.Printf("draining original infrastructure nodes")
+
+	originalNodes := &corev1.NodeList{}
+	if err := r.client.List(ctx, originalNodes, &client.ListOptions{LabelSelector: originalNodeSelector}); err != nil {
+		return err
+	}
+
+	for _, node := range originalNodes.Items {
+
+		node.Annotations[drainingInProgressAnnotation] = "true"
+		node.Annotations[replacementInProgressAnnotation] = "true"
+
+		// Unschedulable + noExecute taint should immediately cause nodes to drain
+		node.Spec.Taints = append(node.Spec.Taints, *unschedulableTaint)
+		node.Spec.Taints = append(node.Spec.Taints, *noExecuteTaint)
+		node.Spec.Unschedulable = true
+
+		if err = r.client.Update(ctx, &node); err != nil {
+			return err
+		}
+	}
```

Comment:
It does, actually.  The "NoExecute" taint immediately evicts pods from the node, and "NoSchedule" prevents them from being reschedule on that node.  You can test it manually by just adding the any two keys, one with the "NoExecute" and one with "NoSchedule" as the effect, to any node. :)

But point taken, I think given the machinepool deletion it's not necessary.  I was thinking it might give us a jump on the draining, but since it ended up removing the pool _immediately_ after this, it's no longer needed.

I'll remove this stuff.

Reply:
> Typically an eviction resource needs to be created https://kubernetes.io/docs/concepts/scheduling-eviction/api-eviction/

Ah, TIL, thanks!

With the two taints, the NodeController is evicting the pods, so I'm guess it's the one creating the eviction resource, but doing so due to the taints added.

Reply:
Ah TIL myself :D https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#taint-based-evictions